### PR TITLE
TAM 2.0 spec proposal.

### PR DESCRIPTION
**Pull request by SarathBug**
SAI TAM 2.0 spec proposal from BRCM.

### CODE REVIEW

File: inc/saitypes.h [hunk 1]
```diff
@@ -244,7 +244,23 @@ typedef enum _sai_object_type_t
...
+    SAI_OBJECT_TYPE_TAM_TEL_INT              = 83,
+    SAI_OBJECT_TYPE_TAM_TRANSPORT            = 84,
+    SAI_OBJECT_TYPE_TAM_IFA_HDR              = 85,
+    SAI_OBJECT_TYPE_TAM_IFA_PROBE            = 86,
+    SAI_OBJECT_TYPE_TAM_PROBE                = 87,
+    SAI_OBJECT_TYPE_TAM_TELEMETRY            = 88,
+    SAI_OBJECT_TYPE_TAM_COLLECTOR            = 89,
+    SAI_OBJECT_TYPE_TAM_EVENT_ACTION         = 90,
+    SAI_OBJECT_TYPE_TAM_EVENT                = 91,
+    SAI_OBJECT_TYPE_MAX                      = 92,
```
Comment (kcudnik) [inline inc/saitypes.h]: why do we need so many object types ? is it possible to keep this list smaller ? and maybe not so many of them need to be objects, but a single attributes value ?

File: inc/saitam.h [hunk 2]
```diff
@@ -868,29 +1012,2440 @@ typedef sai_status_t (*sai_get_tam_transporter_attribute_fn)(
...
+
+    /**
+     * @brief Start of Attributes
+     */
+    SAI_TAM_EVENT_THRESHOLD_ATTR_START,
+
+    /**
+     * @brief High water mark %
+     *
+     * @type sai_uint32_t
```
Comment (zhenggen-xu) [inline inc/saitam.h]: Is this value between 0-100?  should we use sai_uint8_t ?

File: inc/saitam.h [hunk 3]
```diff
@@ -868,29 +1012,2440 @@ typedef sai_status_t (*sai_get_tam_transporter_attribute_fn)(
...
+     */
+    SAI_TAM_TELEMETRY_TYPE_FABRIC,
+
+    /**
+     * @brief Flow TAM
+     */
+    SAI_TAM_TELEMETRY_TYPE_FLOW,
+
+    /**
+     * @brief Networking element TAM
```
Comment (zhenggen-xu) [inline inc/saitam.h]: INT TAM?

File: meta/acronyms.txt [hunk 4]
```diff
@@ -50,22 +54,25 @@ PBS - Peak Burst Size
 PCI - Peripheral Component Interconnect
 PFC - Priority Flow Control
 PIR - Peak Information Rate
-QOS - Quality of Service
```
Comment (zhenggen-xu) [inline meta/acronyms.txt]: reason why this is removed?

File: inc/saitam.h [hunk 5]
```diff
@@ -868,29 +1012,2440 @@ typedef sai_status_t (*sai_get_tam_transporter_attribute_fn)(
...
+    /** Custom range base value */
+    SAI_TAM_MATH_FUNC_ATTR_CUSTOM_RANGE_START = 0x10000000,
+
+    /** End of custom range base */
+    SAI_TAM_MATH_FUNC_ATTR_CUSTOM_RANGE_END
+
+} sai_tam_math_func_attr_t;
+
+/**
+ * @brief Create and return a match function object
```
Comment (zhenggen-xu) [inline inc/saitam.h]: math instead of match?

File: inc/saitam.h [hunk 6]
```diff
@@ -868,29 +1012,2440 @@ typedef sai_status_t (*sai_get_tam_transporter_attribute_fn)(
...
+ * @return #SAI_STATUS_SUCCESS on success, failure status code on error
+ */
+typedef sai_status_t (*sai_create_tam_math_func_fn)(
+        _Out_ sai_object_id_t *tam_math_func_id,
+        _In_ sai_object_id_t switch_id,
+        _In_ uint32_t attr_count,
+        _In_ const sai_attribute_t *attr_list);
+
+/**
+ * @brief Deletes a specified Match function object
```
Comment (zhenggen-xu) [inline inc/saitam.h]: math instead of match?

File: inc/saitam.h (hunk 5 above)
Comment (SarathBug) [reply inc/saitam.h]: will fix it

File: inc/saitam.h (hunk 6 above)
Comment (SarathBug) [reply inc/saitam.h]: will fix it

File: inc/saitam.h (hunk 2 above)
Comment (SarathBug) [reply inc/saitam.h]: will change it

File: meta/acronyms.txt (hunk 4 above)
Comment (SarathBug) [reply meta/acronyms.txt]: metadata checker was complaining about QoS/qos in  saitam.h. I do not see it any more. Will add it back

File: inc/saitypes.h (hunk 1 above)
Comment (SarathBug) [reply inc/saitypes.h]: Reduced SAI_OBJECT_TYPE_TAM_TEL_XXXXX to one SAI_OBJECT_TYPE_TAM_TEL_TYPE. The other object types have different applications.  I am not sure whether they can be aggregated.

**Comment by kcudnik** [comment]
can you resolve conflicts ?

File: inc/saitam.h [hunk 7]
```diff
@@ -868,29 +1012,2440 @@ typedef sai_status_t (*sai_get_tam_transporter_attribute_fn)(
...
+     * @brief INT template Id
+     *
+     * Template id indicates the supported metadata header.
+     * Device can support one or more template IDs
+     *
+     * @type sai_uint8_t
+     * @flags CREATE_AND_SET
+     * @default 0
+     */
+    SAI_TAM_TEL_INT_ATTR_TEMPLATE_ID = SAI_TAM_TEL_INT_ATTR_START,
```
Comment (dproc) [inline inc/saitam.h]: Where is the "template" defined. Not found in this spec

File: inc/saitam.h [hunk 8]
```diff
@@ -868,29 +1012,2440 @@ typedef sai_status_t (*sai_get_tam_transporter_attribute_fn)(
...
+    /**
+     * @brief Math func attached
+     *
+     * @type sai_object_id_t
+     * @flags CREATE_AND_SET
+     * @objects SAI_OBJECT_TYPE_TAM_MATH_FUNC
+     * @allownull true
+     * @default SAI_NULL_OBJECT_ID
+     */
+    SAI_TAM_TEL_INT_ATTR_MATH_FUNC,
```
Comment (dproc) [inline inc/saitam.h]: what is the math func for? Is it specifying how to process the metadata populated into the INT header?

File: inc/saitam.h [hunk 9]
```diff
@@ -868,29 +1012,2440 @@ typedef sai_status_t (*sai_get_tam_transporter_attribute_fn)(
...
+ * @brief Create and return a telemetry int object
+ *
+ * @param[out] tam_tel_int_id Telemetry int object
+ * @param[in] switch_id Switch object id
+ * @param[in] attr_count Number of attributes
+ * @param[in] attr_list Array of attributes
+ *
+ * @return #SAI_STATUS_SUCCESS on success, failure status code on error
+ */
+typedef sai_status_t (*sai_create_tam_tel_int_fn)(
```
Comment (dproc) [inline inc/saitam.h]: Where can user configure watchlist to specify flows of interest?

File: inc/saitam.h (hunk 7 above)
Comment (SarathBug) [reply inc/saitam.h]: Template Id as specified in https://tools.ietf.org/html/draft-kumar-ifa-00

**Comment by VitalyVovnoboy** [comment]
A description is not clear and would be good to add info of following fields and/or attributes in saitam.h:
- SAI_TAM_EVENT_TYPE_FLOW_STATE
- SAI_TAM_EVENT_TYPE_FLOW_WATCHLIST
- SAI_TAM_EVENT_TYPE_FLOW_TCPFLAG
Thanks!

**Comment by SarathBug** [comment]
[TAM-API-v2.0.pdf](https://github.com/opencomputeproject/SAI/files/2102595/TAM-API-v2.0.pdf)

TAMv2 API spec

**Pull request closed without merging**